### To do

1. Extract data from xlsx file and create raw data frame - **DONE**
2. Remove empty columns - **DONE**
3. Remove unecessary columns - **DONE**
   1. Check for distinct data values - **DONE**
4. Rename columns - **DONE**
5. Reorder columns - **DONE**
1. Check for empty data - **DONE**
2. Verify data types - **DONE**
3. Create new dataframes by dimensions, indicators, and subgroups.

In [92]:
# Import necessary libraries
import pandas as pd

### Extract data from the data.xlsx file

In [93]:
# Read the data into a Pandas DataFrame
raw_data_df = pd.read_excel('data/data.xlsx')
raw_data_df.head()

setting  date source indicator_abbr  \
0  Afghanistan  2004    NNS     overweight   
1  Afghanistan  2004    NNS     overweight   
2  Afghanistan  2004    NNS     overweight   
3  Afghanistan  2004    NNS     overweight   
4  Afghanistan  2004    NNS     overweight   

                                      indicator_name  \
0  Overweight prevalence in children aged < 5 yea...   
1  Overweight prevalence in children aged < 5 yea...   
2  Overweight prevalence in children aged < 5 yea...   
3  Overweight prevalence in children aged < 5 yea...   
4  Overweight prevalence in children aged < 5 yea...   

                        dimension      subgroup  estimate  se  ci_lb  ...  \
0  Child's age (6 groups) (0-59m)    0-5 months       NaN NaN    NaN  ...   
1  Child's age (6 groups) (0-59m)  12-23 months       4.3 NaN    NaN  ...   
2  Child's age (6 groups) (0-59m)  24-35 months       3.0 NaN    NaN  ...   
3  Child's age (6 groups) (0-59m)  36-47 months       5.6 NaN    NaN  ...   
4  Child's age (6 groups) (0-59m)  48-59 months       6.4 NaN    NaN  ...   

   iso3  favourable_indicator indicator_scale  ordered_dimension  \
0   AFG                     0             100                  1   
1   AFG                     0             100                  1   
2   AFG                     0             100                  1   
3   AFG                     0             100                  1   
4   AFG                     0             100                  1   

  subgroup_order  reference_subgroup                whoreg6  wbincome2024  \
0              1                   0  Eastern Mediterranean    Low income   
1              3                   0  Eastern Mediterranean    Low income   
2              4                   0  Eastern Mediterranean    Low income   
3              5                   0  Eastern Mediterranean    Low income   
4              6                   0  Eastern Mediterranean    Low income   

   dataset_id        update  
0     rep_nut  17 June 2024  
1     rep_nut  17 June 2024  
2     rep_nut  17 June 2024  
3     rep_nut  17 June 2024  
4     rep_nut  17 June 2024  

[5 rows x 24 columns]

In [94]:
# Get a brief summary of the raw_data DataFrame.
raw_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157095 entries, 0 to 157094
Data columns (total 24 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   setting               157095 non-null  object 
 1   date                  157095 non-null  int64  
 2   source                157095 non-null  object 
 3   indicator_abbr        157095 non-null  object 
 4   indicator_name        157095 non-null  object 
 5   dimension             157095 non-null  object 
 6   subgroup              157095 non-null  object 
 7   estimate              153334 non-null  float64
 8   se                    136679 non-null  float64
 9   ci_lb                 142754 non-null  float64
 10  ci_ub                 142754 non-null  float64
 11  population            147089 non-null  float64
 12  flag                  157095 non-null  object 
 13  setting_average       157095 non-null  float64
 14  iso3                  157095 non-null  object 
 15  

There are 157,095 records and 24 columns of data in our dataframe.

Check for any columns that are empty and drop them from the dataframe.

In [95]:
# Check for empty columns
empty_columns = raw_data_df.columns[raw_data_df.isnull().all()]
print("Empty columns:", empty_columns)

# Remove empty columns
cleaned_df = raw_data_df.dropna(axis=1, how='all')

# Verify data frame info
print("DataFrame info after removing empty columns:")
cleaned_df.info()

Empty columns: Index([], dtype='object')
DataFrame info after removing empty columns:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157095 entries, 0 to 157094
Data columns (total 24 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   setting               157095 non-null  object 
 1   date                  157095 non-null  int64  
 2   source                157095 non-null  object 
 3   indicator_abbr        157095 non-null  object 
 4   indicator_name        157095 non-null  object 
 5   dimension             157095 non-null  object 
 6   subgroup              157095 non-null  object 
 7   estimate              153334 non-null  float64
 8   se                    136679 non-null  float64
 9   ci_lb                 142754 non-null  float64
 10  ci_ub                 142754 non-null  float64
 11  population            147089 non-null  float64
 12  flag                  157095 non-null  object 
 13  setting_average   

There are still 24 columns, so no empty columns were found.

Count the unique values for each column so that we can determine if any have a single value, we can drop them from the dataframe.  Such columns will not provide any meaningful insights.

In [96]:
# Count the number of unique values in each column
unique_counts = cleaned_df.nunique()

# Create a new DataFrame to store the counts
unique_counts_df = pd.DataFrame(unique_counts, columns=['Unique Count'])

# Display the new DataFrame
print("DataFrame with the count of unique values in each column:")
unique_counts_df

DataFrame with the count of unique values in each column:


Unique Count
setting                        156
date                            34
source                          19
indicator_abbr                  15
indicator_name                  15
dimension                        6
subgroup                      4517
estimate                    130841
se                          135364
ci_lb                       136109
ci_ub                       136147
population                   73542
flag                          1496
setting_average              11905
iso3                           156
favourable_indicator             1
indicator_scale                  1
ordered_dimension                2
subgroup_order                   7
reference_subgroup               2
whoreg6                          6
wbincome2024                     4
dataset_id                       1
update                           1

Columns 'favourable indicator', 'indicator_scale', 'dataset_id', and 'update' have single values so we can drop them.

In [97]:
# Remove columns with a single value from the DataFrame
cleaned_df = cleaned_df.drop(columns=['favourable_indicator', 'indicator_scale', 'dataset_id', 'update'])

# Display the updated DataFrame
print("DataFrame info after removing single value columns:\n")
cleaned_df.info()

DataFrame info after removing single value columns:



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157095 entries, 0 to 157094
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   setting             157095 non-null  object 
 1   date                157095 non-null  int64  
 2   source              157095 non-null  object 
 3   indicator_abbr      157095 non-null  object 
 4   indicator_name      157095 non-null  object 
 5   dimension           157095 non-null  object 
 6   subgroup            157095 non-null  object 
 7   estimate            153334 non-null  float64
 8   se                  136679 non-null  float64
 9   ci_lb               142754 non-null  float64
 10  ci_ub               142754 non-null  float64
 11  population          147089 non-null  float64
 12  flag                157095 non-null  object 
 13  setting_average     157095 non-null  float64
 14  iso3                157095 non-null  object 
 15  ordered_dimension   157095 non-nul

The four single value columns were correctly dropped.  We now have 20 columns in the dataframe.

In [98]:
'''
***************
DELETE LATER
***************
# Get the unique values for each column in separate lists so that we can determine if any columns have unnecessary data.  
# Such columns will not be applicable to our analysis, so we will drop them from the dataframe.

# Create an empty dictionary to store the lists of unique values
unique_values = {}

# Iterate through each column in the cleaned DataFrame
for column in cleaned_df.columns:
    # Extract unique values and convert to a list
    unique_values[column] = cleaned_df[column].unique().tolist()

# Display the lists of unique values for each column
print("Unique values in each column:\n")
for column, values in unique_values.items():
    print(f"{column}: {values}")
'''

'\n***************\nDELETE LATER\n***************\n# Get the unique values for each column in separate lists so that we can determine if any columns have unnecessary data.  \n# Such columns will not be applicable to our analysis, so we will drop them from the dataframe.\n\n# Create an empty dictionary to store the lists of unique values\nunique_values = {}\n\n# Iterate through each column in the cleaned DataFrame\nfor column in cleaned_df.columns:\n    # Extract unique values and convert to a list\n    unique_values[column] = cleaned_df[column].unique().tolist()\n\n# Display the lists of unique values for each column\nprint("Unique values in each column:\n")\nfor column, values in unique_values.items():\n    print(f"{column}: {values}")\n'

1. The *flag* column includes notes, author, and reference title.
2. The *source* column includes a code for reprenting the source data type.
3. The meaning of the *reference_subgroup* column is unknown.
4. The *se* column refers to the standard error in the prevalence estimate.
5. The *ci_lb* and *ci_ub* columns refer to confidence intervals upper and lower bounds respectively.
6. The *ordered_dimension* column is a boolean flag that indicates if the *dimension* column is a nominal or ordinal variable.  In this dataset, *0* represents a nominal variable and *1* represents an ordinal variable.  Nominal variables are for the dimensions of *sex*, *place of residence*, and *subnational region*.  Ordinal variables are for the dimensions of *age*, *economic status*, and *education level*. Ordinal dimensions use the *subgroup_order* column to denote the specific ordering.

None data is needed for our analysis, so we will drop it from the dataframe.

In [99]:
# Remove 'flag' column from the DataFrame
cleaned_df = cleaned_df.drop(columns=['flag', 'source', 'reference_subgroup', 'se', 'ci_lb', 'ci_ub', 'ordered_dimension'])

# Display the updated DataFrame
print("DataFrame info after removing the unecessary columns:\n")
cleaned_df.info()

DataFrame info after removing the unecessary columns:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157095 entries, 0 to 157094
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   setting          157095 non-null  object 
 1   date             157095 non-null  int64  
 2   indicator_abbr   157095 non-null  object 
 3   indicator_name   157095 non-null  object 
 4   dimension        157095 non-null  object 
 5   subgroup         157095 non-null  object 
 6   estimate         153334 non-null  float64
 7   population       147089 non-null  float64
 8   setting_average  157095 non-null  float64
 9   iso3             157095 non-null  object 
 10  subgroup_order   157095 non-null  int64  
 11  whoreg6          157095 non-null  object 
 12  wbincome2024     157095 non-null  object 
dtypes: float64(3), int64(2), object(8)
memory usage: 15.6+ MB


1. The *population* column represents the number of children under five years of age.
2. The *whoreg6* column is a regional classification provided by the WHO.
3. The *wbincome2024* column is an income group classification provided by The World Bank.

Rename the columns with more intuitive titles and reorder them.

In [101]:
# Rename the columns
cleaned_df = cleaned_df.rename(columns={
    'setting': 'Country',
    'date': 'Year',
    'indicator_abbr': 'Anthropometric Indicator',
    'indicator_name': 'Indicator Description',
    'dimension': 'Dimension',
    'subgroup': 'Dimension Value',
    'subgroup_order': 'Dimension Value Order',
    'setting_average': 'Country Avg',
    'iso3': 'Country ISO-3 Code',
    'whoreg6': 'Region',
    'wbincome2024': 'Income Group',
    'estimate': 'Prevalence Estimate %',
    'population': 'Under-Five Population (million)'
})

# Reorder the columns
cleaned_df = cleaned_df[
    [
        'Region',
        'Country ISO-3 Code',
        'Country',
        'Year',
        'Dimension',
        'Dimension Value',
        'Dimension Value Order',
        'Income Group',
        'Anthropometric Indicator',
        'Indicator Description',
        'Prevalence Estimate %',
        'Under-Five Population (million)',
        'Country Avg'
    ]
]

# Check the DataFrame information
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157095 entries, 0 to 157094
Data columns (total 13 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   Region                           157095 non-null  object 
 1   Country ISO-3 Code               157095 non-null  object 
 2   Country                          157095 non-null  object 
 3   Year                             157095 non-null  int64  
 4   Dimension                        157095 non-null  object 
 5   Dimension Value                  157095 non-null  object 
 6   Dimension Value Order            157095 non-null  int64  
 7   Income Group                     157095 non-null  object 
 8   Anthropometric Indicator         157095 non-null  object 
 9   Indicator Description            157095 non-null  object 
 10  Prevalence Estimate %            153334 non-null  float64
 11  Under-Five Population (million)  147089 non-null  float64
 12  Co

In [102]:
# Remove rows where 'Under-Five Population (million)' and 'Prevalence Estimate %' are empty
cleaned_df = cleaned_df.dropna(subset=['Under-Five Population (million)', 'Prevalence Estimate %'])

# Check the DataFrame after dropping rows
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 147089 entries, 1 to 157094
Data columns (total 13 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   Region                           147089 non-null  object 
 1   Country ISO-3 Code               147089 non-null  object 
 2   Country                          147089 non-null  object 
 3   Year                             147089 non-null  int64  
 4   Dimension                        147089 non-null  object 
 5   Dimension Value                  147089 non-null  object 
 6   Dimension Value Order            147089 non-null  int64  
 7   Income Group                     147089 non-null  object 
 8   Anthropometric Indicator         147089 non-null  object 
 9   Indicator Description            147089 non-null  object 
 10  Prevalence Estimate %            147089 non-null  float64
 11  Under-Five Population (million)  147089 non-null  float64
 12  Country

After removing unecessary columns and empty rows, we are left with 147,089 entries and 13 columns.